### PR TITLE
[Merged by Bors] - feat(CategoryTheory/SmallObject): the category of transfinite iterations of a functor

### PR DESCRIPTION
**Pull request by joelriou**
In this PR, we construct the category of transfinite iterations of a functor. We show that there is at most one morphism between two objects in this category. In future PRs, we shall show (by transfinite induction) that there is exactly one (iso)morphism between two objects, and that there is always an object (that is unique up to a unique isomorphism).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/SmallObject/Iteration.lean [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+/-- Given `F : { i // i ≤ j } ⥤ C`, `i : J` such that `hi : i ≤ j`, this is the
+cocone consisting of all maps `F.obj ⟨k, hk⟩ ⟶ F.obj ⟨i, hi⟩` for `k : J` such that `k < i`. -/
+@[simps]
+def coconeOfLE : Cocone (restrictionLT F hi) where
+  pt := F.obj ⟨i, hi⟩
+  ι :=
+    { app := fun ⟨k, hk⟩ => F.map (homOfLE (by simpa using hk.le))
+      naturality := fun ⟨k₁, hk₁⟩ ⟨k₂, hk₂⟩ _ => by
+        dsimp
+        rw [comp_id, ← Functor.map_comp, homOfLE_comp] }
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/SmallObject/Iteration.lean:78]:
```suggestion
        simp [comp_id, ← Functor.map_comp, homOfLE_comp] }
```

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/CategoryTheory/SmallObject/Iteration.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/SmallObject/Iteration.lean:78]: Thanks!

**Review by alreadydone — COMMENTED** [review]
Interesting that there's so much API overlap with my #9480 ...

**Comment by joelriou** [comment]
> Interesting that there's so much API overlap with my #9480 ...

I originally did this about one and half year ago in Lean 3 https://github.com/joelriou/homotopical_algebra/blob/main/src/for_mathlib/algebraic_topology/small_object/iteration.lean
I guess that this is what we need to do when we need to construct data by transfinite induction...

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**